{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "85bb38b1-f1ad-49ab-a6d2-9d247787cec4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np \n",
    "import pandas as pd \n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8e0aaf8e-d155-4e2d-9d75-0453f382538a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Gene             ENSP        Variation      VAR             ENST Mutation  \\\n",
      "0   RAF1  ENSP00000251849              NaN  VAR1429  ENST00000251849    P261T   \n",
      "1   HRAS  ENST00000311189  ENSP00000309845  VAR0775              NaN     G12D   \n",
      "2   PTEN  ENSP00000361021              NaN  VAR0714  ENST00000371953    P204A   \n",
      "3   KRAS              NaN  ENSP00000308495  VAR0966  ENST00000311936    F156I   \n",
      "4  STAT3  ENSP00000264657              NaN  VAR1222  ENST00000264657    R382Q   \n",
      "\n",
      "  Ref AA  Position Final AA  Randomized  ... PredictSNP  \\\n",
      "0      P       261        T    0.548635  ...  -0.718713   \n",
      "1      G        12        D    0.293770  ...  -0.869084   \n",
      "2      P       204        A    0.868490  ...  -0.654946   \n",
      "3      F       156        I    0.675611  ...  -0.869084   \n",
      "4      R       382        Q    0.754193  ...  -0.755661   \n",
      "\n",
      "   PredictSNP Prediction MutPred  MutPred Prediction   VEST  Vest Prediction  \\\n",
      "0            Deleterious   0.595         Deleterious  0.785      Deleterious   \n",
      "1            Deleterious   0.932         Deleterious  0.836      Deleterious   \n",
      "2            Deleterious   0.732         Deleterious  0.856      Deleterious   \n",
      "3            Deleterious   0.923         Deleterious  0.868      Deleterious   \n",
      "4            Deleterious   0.716         Deleterious  0.935      Deleterious   \n",
      "\n",
      "  VEST FDR  Vest FDR Prediction VEST P  Vest P Prediction  \n",
      "0   0.7272          Deleterious  0.808        Deleterious  \n",
      "1   0.8360          Deleterious  0.836        Deleterious  \n",
      "2   0.8560          Deleterious  0.856        Deleterious  \n",
      "3   0.8680          Deleterious  0.868        Deleterious  \n",
      "4   0.9350          Deleterious  0.935        Deleterious  \n",
      "\n",
      "[5 rows x 23 columns]\n"
     ]
    }
   ],
   "source": [
    "#file MOESM_3\n",
    "MOESM_3 = '41598_2024_71422_MOESM3_ESM.xlsx'\n",
    "df = pd.read_excel(MOESM_3)\n",
    "print(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9fad5374-72b3-4efc-9099-941f4d75e8e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Elapsed: 28.08272779599065\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>ENSP</th>\n",
       "      <th>Variation</th>\n",
       "      <th>VAR</th>\n",
       "      <th>ENST</th>\n",
       "      <th>Position</th>\n",
       "      <th>Ref AA</th>\n",
       "      <th>Final AA</th>\n",
       "      <th>wild_seq</th>\n",
       "      <th>mut_seq</th>\n",
       "      <th>error</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RAF1</td>\n",
       "      <td>ENSP00000251849</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VAR1429</td>\n",
       "      <td>ENST00000251849</td>\n",
       "      <td>261</td>\n",
       "      <td>P</td>\n",
       "      <td>T</td>\n",
       "      <td>MEHIQGAWKTISNGFGFKDAVFDGSSCISPTIVQQFGYQRRASDDG...</td>\n",
       "      <td>MEHIQGAWKTISNGFGFKDAVFDGSSCISPTIVQQFGYQRRASDDG...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HRAS</td>\n",
       "      <td>ENST00000311189</td>\n",
       "      <td>ENSP00000309845</td>\n",
       "      <td>VAR0775</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>G</td>\n",
       "      <td>D</td>\n",
       "      <td>AGGCCCGCCCGAGTCTCCGCCGCCCGTGCCCTGCGCCCGCAACCCG...</td>\n",
       "      <td>None</td>\n",
       "      <td>fetch/mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PTEN</td>\n",
       "      <td>ENSP00000361021</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VAR0714</td>\n",
       "      <td>ENST00000371953</td>\n",
       "      <td>204</td>\n",
       "      <td>P</td>\n",
       "      <td>A</td>\n",
       "      <td>MTAIIKEIVSRNKRRYQEDGFDLDLTYIYPNIIAMGFPAERLEGVY...</td>\n",
       "      <td>MTAIIKEIVSRNKRRYQEDGFDLDLTYIYPNIIAMGFPAERLEGVY...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000308495</td>\n",
       "      <td>VAR0966</td>\n",
       "      <td>ENST00000311936</td>\n",
       "      <td>156</td>\n",
       "      <td>F</td>\n",
       "      <td>I</td>\n",
       "      <td>LGGGRGGGGSSGGGSGGGEGGGGSASTPGPRHFGLGASAAQALKAA...</td>\n",
       "      <td>None</td>\n",
       "      <td>fetch/mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STAT3</td>\n",
       "      <td>ENSP00000264657</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VAR1222</td>\n",
       "      <td>ENST00000264657</td>\n",
       "      <td>382</td>\n",
       "      <td>R</td>\n",
       "      <td>Q</td>\n",
       "      <td>MAQWNQLQQLDTRYLEQLHQLYSDSFPMELRQFLAPWIESQDWAYA...</td>\n",
       "      <td>MAQWNQLQQLDTRYLEQLHQLYSDSFPMELRQFLAPWIESQDWAYA...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>ASH1L</td>\n",
       "      <td>ENSP00000376204</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VAR0593</td>\n",
       "      <td>ENST00000392403</td>\n",
       "      <td>1081</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>MDPRNTAMLGLGSDSEGFSRKSPSAISTGTLVSKREVELEKNTKEE...</td>\n",
       "      <td>MDPRNTAMLGLGSDSEGFSRKSPSAISTGTLVSKREVELEKNTKEE...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>ASXL2</td>\n",
       "      <td>ENSP00000391447</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VAR0522</td>\n",
       "      <td>ENST00000435504</td>\n",
       "      <td>1034</td>\n",
       "      <td>R</td>\n",
       "      <td>S</td>\n",
       "      <td>MREKGRRKKGRTWAEAAKTVLEKYPNTPMSHKEILQVIQREGLKEI...</td>\n",
       "      <td>MREKGRRKKGRTWAEAAKTVLEKYPNTPMSHKEILQVIQREGLKEI...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>ENST00000380152</td>\n",
       "      <td>ENSP00000369497</td>\n",
       "      <td>VAR0472</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1396</td>\n",
       "      <td>Q</td>\n",
       "      <td>R</td>\n",
       "      <td>AGAGGCGGAGCCGCTGTGGCACTGCTGCGCCTCTGCTGCGCCTCGG...</td>\n",
       "      <td>None</td>\n",
       "      <td>fetch/mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>HRAS</td>\n",
       "      <td>ENST00000311189</td>\n",
       "      <td>ENSP00000309845</td>\n",
       "      <td>VAR0569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>86</td>\n",
       "      <td>N</td>\n",
       "      <td>T</td>\n",
       "      <td>AGGCCCGCCCGAGTCTCCGCCGCCCGTGCCCTGCGCCCGCAACCCG...</td>\n",
       "      <td>None</td>\n",
       "      <td>fetch/mismatch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>ENST00000380152</td>\n",
       "      <td>ENSP00000369497</td>\n",
       "      <td>VAR0344</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3292</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>AGAGGCGGAGCCGCTGTGGCACTGCTGCGCCTCTGCTGCGCCTCGG...</td>\n",
       "      <td>None</td>\n",
       "      <td>fetch/mismatch</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene             ENSP        Variation      VAR             ENST  \\\n",
       "0     RAF1  ENSP00000251849              NaN  VAR1429  ENST00000251849   \n",
       "1     HRAS  ENST00000311189  ENSP00000309845  VAR0775              NaN   \n",
       "2     PTEN  ENSP00000361021              NaN  VAR0714  ENST00000371953   \n",
       "3     KRAS              NaN  ENSP00000308495  VAR0966  ENST00000311936   \n",
       "4    STAT3  ENSP00000264657              NaN  VAR1222  ENST00000264657   \n",
       "..     ...              ...              ...      ...              ...   \n",
       "195  ASH1L  ENSP00000376204              NaN  VAR0593  ENST00000392403   \n",
       "196  ASXL2  ENSP00000391447              NaN  VAR0522  ENST00000435504   \n",
       "197  BRCA2  ENST00000380152  ENSP00000369497  VAR0472              NaN   \n",
       "198   HRAS  ENST00000311189  ENSP00000309845  VAR0569              NaN   \n",
       "199  BRCA2  ENST00000380152  ENSP00000369497  VAR0344              NaN   \n",
       "\n",
       "     Position Ref AA Final AA  \\\n",
       "0         261      P        T   \n",
       "1          12      G        D   \n",
       "2         204      P        A   \n",
       "3         156      F        I   \n",
       "4         382      R        Q   \n",
       "..        ...    ...      ...   \n",
       "195      1081      A        T   \n",
       "196      1034      R        S   \n",
       "197      1396      Q        R   \n",
       "198        86      N        T   \n",
       "199      3292      P        L   \n",
       "\n",
       "                                              wild_seq  \\\n",
       "0    MEHIQGAWKTISNGFGFKDAVFDGSSCISPTIVQQFGYQRRASDDG...   \n",
       "1    AGGCCCGCCCGAGTCTCCGCCGCCCGTGCCCTGCGCCCGCAACCCG...   \n",
       "2    MTAIIKEIVSRNKRRYQEDGFDLDLTYIYPNIIAMGFPAERLEGVY...   \n",
       "3    LGGGRGGGGSSGGGSGGGEGGGGSASTPGPRHFGLGASAAQALKAA...   \n",
       "4    MAQWNQLQQLDTRYLEQLHQLYSDSFPMELRQFLAPWIESQDWAYA...   \n",
       "..                                                 ...   \n",
       "195  MDPRNTAMLGLGSDSEGFSRKSPSAISTGTLVSKREVELEKNTKEE...   \n",
       "196  MREKGRRKKGRTWAEAAKTVLEKYPNTPMSHKEILQVIQREGLKEI...   \n",
       "197  AGAGGCGGAGCCGCTGTGGCACTGCTGCGCCTCTGCTGCGCCTCGG...   \n",
       "198  AGGCCCGCCCGAGTCTCCGCCGCCCGTGCCCTGCGCCCGCAACCCG...   \n",
       "199  AGAGGCGGAGCCGCTGTGGCACTGCTGCGCCTCTGCTGCGCCTCGG...   \n",
       "\n",
       "                                               mut_seq           error  \n",
       "0    MEHIQGAWKTISNGFGFKDAVFDGSSCISPTIVQQFGYQRRASDDG...            None  \n",
       "1                                                 None  fetch/mismatch  \n",
       "2    MTAIIKEIVSRNKRRYQEDGFDLDLTYIYPNIIAMGFPAERLEGVY...            None  \n",
       "3                                                 None  fetch/mismatch  \n",
       "4    MAQWNQLQQLDTRYLEQLHQLYSDSFPMELRQFLAPWIESQDWAYA...            None  \n",
       "..                                                 ...             ...  \n",
       "195  MDPRNTAMLGLGSDSEGFSRKSPSAISTGTLVSKREVELEKNTKEE...            None  \n",
       "196  MREKGRRKKGRTWAEAAKTVLEKYPNTPMSHKEILQVIQREGLKEI...            None  \n",
       "197                                               None  fetch/mismatch  \n",
       "198                                               None  fetch/mismatch  \n",
       "199                                               None  fetch/mismatch  \n",
       "\n",
       "[200 rows x 11 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import time\n",
    "import pandas as pd\n",
    "import requests\n",
    "from multiprocess import Pool\n",
    "from time import sleep\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "# ——— Setup a shared session ———\n",
    "session = requests.Session()\n",
    "session.headers.update({\"Content-Type\": \"application/json\", \"Accept\": \"application/json\"})\n",
    "\n",
    "# ——— 1) Fetch protein by ENSP ———\n",
    "def fetch_ensembl_protein_sequence(ensp_id, retry=2, pause=0.1):\n",
    "    url = f\"https://rest.ensembl.org/sequence/id/{ensp_id}\"\n",
    "    headers = {\"Content-Type\": \"text/plain\"}\n",
    "    for _ in range(retry):\n",
    "        r = session.get(url, headers=headers, timeout=5)\n",
    "        if r.status_code == 200:\n",
    "            return r.text.replace(\"\\n\", \"\")\n",
    "        sleep(pause)\n",
    "    return None\n",
    "\n",
    "# ——— 2) Fetch cDNA by ENST and translate ———\n",
    "def translate_cdna_from_enst(enst_id, retry=2, pause=0.1):\n",
    "    url = f\"https://rest.ensembl.org/sequence/id/{enst_id}?type=cdna\"\n",
    "    headers = {\"Content-Type\": \"text/plain\"}\n",
    "    for _ in range(retry):\n",
    "        r = session.get(url, headers=headers, timeout=5)\n",
    "        if r.status_code == 200:\n",
    "            cdna = r.text.replace(\"\\n\", \"\")\n",
    "            # Trim to multiple of 3\n",
    "            rem = len(cdna) % 3\n",
    "            if rem:\n",
    "                cdna = cdna[:-rem]\n",
    "            protein = str(Seq(cdna).translate(to_stop=True))\n",
    "            return protein\n",
    "        sleep(pause)\n",
    "    return None\n",
    "\n",
    "\n",
    "# ——— 3) Fetch variation metadata by VAR → canonical transcript/protein ———\n",
    "def fetch_variation_metadata(var_id):\n",
    "    url = f\"https://rest.ensembl.org/variation/human/{var_id}\"\n",
    "    r = session.get(url, timeout=5)\n",
    "    if r.status_code == 200:\n",
    "        return r.json()\n",
    "    return {}\n",
    "\n",
    "def select_canonical_from_variation(var_info):\n",
    "    # pick the first transcript consequence marked canonical\n",
    "    for tc in var_info.get(\"transcript_consequences\", []):\n",
    "        if tc.get(\"is_canonical\"):\n",
    "            return tc.get(\"protein_id\") or tc.get(\"transcript_id\")\n",
    "    # fallback to any first transcript\n",
    "    tc0 = var_info.get(\"transcript_consequences\", [{}])[0]\n",
    "    return tc0.get(\"protein_id\") or tc0.get(\"transcript_id\")\n",
    "\n",
    "# ——— 4) Gene → ENSG → canonical transcript → translate ———\n",
    "def map_symbol_to_ensg(gene_symbol):\n",
    "    url = f\"https://rest.ensembl.org/xrefs/symbol/human/{gene_symbol}\"\n",
    "    r = session.get(url, timeout=5)\n",
    "    if r.status_code == 200:\n",
    "        for entry in r.json():\n",
    "            if entry.get(\"type\") == \"gene\":\n",
    "                return entry[\"id\"]\n",
    "    return None\n",
    "\n",
    "def select_canonical_transcript_by_gene(ensg):\n",
    "    url = f\"https://rest.ensembl.org/lookup/id/{ensg}?expand=1\"\n",
    "    r = session.get(url, timeout=5)\n",
    "    if r.status_code == 200:\n",
    "        for t in r.json().get(\"Transcript\", []):\n",
    "            if t.get(\"is_canonical\"):\n",
    "                return t[\"id\"]\n",
    "    return None\n",
    "\n",
    "# ——— Master retrieval logic ———\n",
    "def retrieve_wt_protein(row):\n",
    "    # 1) ENSP\n",
    "    if pd.notna(row.get(\"ENSP\")):\n",
    "        seq = fetch_ensembl_protein_sequence(row[\"ENSP\"])\n",
    "        if seq: return seq\n",
    "\n",
    "    # 2) ENST\n",
    "    if pd.notna(row.get(\"ENST\")):\n",
    "        seq = translate_cdna_from_enst(row[\"ENST\"])\n",
    "        if seq: return seq\n",
    "\n",
    "    # 3) VAR\n",
    "    if pd.notna(row.get(\"VAR\")):\n",
    "        info = fetch_variation_metadata(row[\"VAR\"])\n",
    "        can_id = select_canonical_from_variation(info)\n",
    "        if can_id and can_id.startswith(\"ENSP\"):\n",
    "            seq = fetch_ensembl_protein_sequence(can_id)\n",
    "        else:\n",
    "            seq = translate_cdna_from_enst(can_id)\n",
    "        if seq: return seq\n",
    "\n",
    "    # 4) Gene\n",
    "    if pd.notna(row.get(\"Gene\")):\n",
    "        ensg = map_symbol_to_ensg(row[\"Gene\"])\n",
    "        if ensg:\n",
    "            tx = select_canonical_transcript_by_gene(ensg)\n",
    "            if tx:\n",
    "                seq = translate_cdna_from_enst(tx)\n",
    "                if seq: return seq\n",
    "\n",
    "    return None\n",
    "\n",
    "# ——— Variant processing ———\n",
    "def process_variant(row):\n",
    "    wt = retrieve_wt_protein(row)\n",
    "    pos, ref, mut = int(row[\"Position\"]), row[\"Ref AA\"], row[\"Final AA\"]\n",
    "    if not wt or pos<1 or pos>len(wt) or wt[pos-1]!=ref:\n",
    "        return {\"wild_seq\": wt, \"mut_seq\": None, \"error\":\"fetch/mismatch\"}\n",
    "    m = list(wt); m[pos-1]=mut\n",
    "    return {\"wild_seq\": wt, \"mut_seq\": \"\".join(m), \"error\":None}\n",
    "\n",
    "# ——— Main ———\n",
    "df = pd.read_excel(\"41598_2024_71422_MOESM3_ESM.xlsx\")\n",
    "cols = [\"Gene\",\"ENSP\",\"Variation\",\"VAR\",\"ENST\",\"Position\",\"Ref AA\",\"Final AA\"]\n",
    "df = df[cols].head(200).copy()\n",
    "\n",
    "t0 = time.perf_counter()\n",
    "with Pool(4) as pool:\n",
    "    res = pool.map(process_variant, df.to_dict(\"records\"))\n",
    "print(\"Elapsed:\", time.perf_counter()-t0)\n",
    "\n",
    "out = pd.DataFrame(res)\n",
    "df[[\"wild_seq\",\"mut_seq\",\"error\"]] = out\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6a8bfe08-57d7-4d38-a6bb-d3f21f18d84e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "86 out of 200 variants have no mut_seq (43.0%).\n"
     ]
    }
   ],
   "source": [
    "# Number of missing mutant sequences\n",
    "missing = df[\"mut_seq\"].isna().sum()\n",
    "total   = len(df)\n",
    "print(f\"{missing} out of {total} variants have no mut_seq \"\n",
    "      f\"({missing/total*100:.1f}%).\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e07075fa-9cfe-446f-8908-e5b904562f80",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (sequence_kernels)",
   "language": "python",
   "name": "sequence_kernels"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
